File: FindIntersection.py
'''
Solution
1.  Find lengths of both the LLs. Take the difference
2.  Move forward the pointer on longest LL upto obtained difference number of times.
3.  Move pointers in both LLs at the same pace and check for intersection.

Time Complexity: O(n) and Space Complexity: O(1)

--- Passed all Testcases on Leetcode successfully
'''

# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class FindIntersection(object):

    def __getLength(self, head):

        '''
        :param head: head of the LL
        :return: length of the LL
        '''

        count = 0

        if (head == None):
            return count

        currNode = head

        while (currNode != None):
            currNode = currNode.next
            count += 1

        return count

    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """

        #   edge case check
        if (headA == None or headB == None):
            return None

        #   obtain lengths using a private method
        countA = self.__getLength(headA)
        countB = self.__getLength(headB)

        #   if first LL is bigger
        if (countA > countB):
            diff = countA - countB
            count = 0

            #   initialize fast pointer to head of LinkedList-A and move it upto the difference
            fastNode = headA
            while (count < diff):
                fastNode = fastNode.next
                count += 1

            #   initialize slow pointer to head of LinkedList-B
            slowNode = headB

            #   move both the pointers until they intersect or reach the end.
            while (fastNode != None):
                if (slowNode == fastNode):
                    return slowNode
                slowNode = slowNode.next
                fastNode = fastNode.next

            return None

        #   if second LL is bigger
        else:
            diff = countB - countA
            count = 0

            #   initialize fast pointer to head of LinkedList-B and move it upto the difference
            fastNode = headB
            while (count < diff):
                fastNode = fastNode.next
                count += 1

            #   initialize slow pointer to head of LinkedList-B
            slowNode = headA

            #   move both the pointers until they intersect or reach the end.
            while (fastNode != None):
                if (slowNode == fastNode):
                    return slowNode
                slowNode = slowNode.next
                fastNode = fastNode.next

            return None

File: findintersection.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

//Time complexity : O(m + n) // where m is length of 1st list and n is the length of 2nd list
//Space complexity : O(1)

//Leetcode : Yes
//Approach :
            // Find length of both the lists
            // Find the difference between both the lengths
            // Whichever has greater length, increment the headpointer till difference is same
            // Once the length is same, start incrementing both the pointers equally
            // Once both the pointers are equal, return any one of the pointer

class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        if(headA == nullptr || headB == nullptr)
        {
            return nullptr;
        }
        
        int l1 = 0;
        int l2 = 0;
        ListNode* ptr = headA;
        while(ptr != nullptr)
        {
            ptr = ptr->next;
            l1++;
        }
        ptr = headB;
        while(ptr != nullptr)
        {
            ptr = ptr->next;
            l2++;
        }
        
       while(l1 > l2)
       {
           headA = headA->next;
           l1--;
       }
       while(l2 > l1)
       {
           headB = headB->next;
           l2--;
       }
        
       while(headA != headB)
       {
           headA = headA->next;
           headB = headB->next;
       }
        
       return headA;
    }
};